app/controllers/api/v0/comments_controller.rb

```diff
@@ -3,21 +3,43 @@ module V0
     class CommentsController < ApiController
       respond_to :json
 
-      caches_action :index,
-                    cache_path: proc { |c| c.params.permit! },
-                    expires_in: 10.minutes
-
-      caches_action :show,
-                    cache_path: proc { |c| c.params.permit! },
-                    expires_in: 10.minutes
+      before_action :set_cache_control_headers, only: %i[index show]
 
       def index
         article = Article.find(params[:a_id])
-        @comments = article.comments.roots.includes(:user).select(%i[id processed_html user_id ancestry])
+
+        @comments = article.comments.includes(:user).select(%i[id processed_html user_id ancestry]).arrange
+
+        set_surrogate_key_header "api_comments_index", edge_cache_keys(@comments)
       end
 
       def show
-        @comment = Comment.find(params[:id].to_i(26))
+        tree_with_root_comment = Comment.subtree_of(params[:id].to_i(26)).
```

Comment:
Given that users are very large objects and we might be loading a lot of them here I tried to look for ways we could limit the user data we were pulling as well and came up with using a `joins` to help limit the data and eliminate a db hit. But I might be missing something here since I have never worked with ancestry so let me know if I am! 

```ruby
[36] pry(main)> Comment.subtree_of(30).joins(:user).select("users.name, users.twitter_username").select(%i[id processed_html user_id ancestry]).arrange
  Comment Load (1.0ms)  SELECT  "comments".* FROM "comments" WHERE "comments"."id" = $1 LIMIT $2  [["id", 30], ["LIMIT", 1]]
  Comment Load (5.8ms)  SELECT users.name, users.twitter_username, "comments"."id", "comments"."processed_html", "comments"."user_id", "comments"."ancestry" FROM "comments" INNER JOIN "users" ON "users"."id" = "comments"."user_id" WHERE (("comments"."ancestry" LIKE '30/%' OR "comments"."ancestry" = '30') OR "comments"."id" = 30)
=> {#<Comment:0x00007f895ced4da0
  id: 30,
  ancestry: nil,
  processed_html: "<!DOCTYPE html PUBLIC \"-//W3C//DTD HTML 4.0 Transitional//EN\" \"http://www.w3.org/TR/REC-html40/loose.dtd\">\n<html><body>\n<p>Cleanse lomo trust fund lumbersexual swag kombucha poutine.</p>\n\n</body></html>\n",
  user_id: 6>=>{}}
[37] pry(main)> Comment.subtree_of(30).includes(:user).select(%i[id processed_html user_id ancestry]).arrange
  Comment Load (0.4ms)  SELECT  "comments".* FROM "comments" WHERE "comments"."id" = $1 LIMIT $2  [["id", 30], ["LIMIT", 1]]
  Comment Load (0.3ms)  SELECT "comments"."id", "comments"."processed_html", "comments"."user_id", "comments"."ancestry" FROM "comments" WHERE (("comments"."ancestry" LIKE '30/%' OR "comments"."ancestry" = '30') OR "comments"."id" = 30)
  User Load (0.2ms)  SELECT "users".* FROM "users" WHERE "users"."id" = $1  [["id", 6]]
=> {#<Comment:0x00007f896ab54578
  id: 30,
  ancestry: nil,
  processed_html: "<!DOCTYPE html PUBLIC \"-//W3C//DTD HTML 4.0 Transitional//EN\" \"http://www.w3.org/TR/REC-html40/loose.dtd\">\n<html><body>\n<p>Cleanse lomo trust fund lumbersexual swag kombucha poutine.</p>\n\n</body></html>\n",
  user_id: 6>=>{}}
[38] pry(main)>
```

Reply:
If we user `joins` there will be additional sql query each time we call `comment.user` (so, N+1)

Reply:
We can't use `joins` or `left_outer_joins`, as @lightalloy said, you get a N+1 each time you access comment.user. This is taken from me calling `curl http://localhost:3000/api/comments\?a_id\=13` where I just left a bunch of threaded comments in an article. The same happens for the `show` action because we load all children comments.

Notice how Rails issues a separate SQL query for each user:

```sql
Processing by Api::V0::CommentsController#index as JSON
  Parameters: {"a_id"=>"13"}
  Article Load (0.5ms)  SELECT  "articles".* FROM "articles" WHERE "articles"."id" = $1 LIMIT $2  [["id", 13], ["LIMIT", 1]]
  Comment Load (0.7ms)  SELECT "comments"."id", "comments"."processed_html", "comments"."user_id", "comments"."ancestry" FROM "comments" INNER JOIN "users" ON "users"."id" = "comments"."user_id" WHERE "comments"."commentable_id" = $1 AND "comments"."commentable_type" = $2  [["commentable_id", 13], ["commentable_type", "Article"]]
  User Load (0.4ms)  SELECT  "users".* FROM "users" WHERE "users"."id" = $1 LIMIT $2  [["id", 3], ["LIMIT", 1]]
  User Load (0.5ms)  SELECT  "users".* FROM "users" WHERE "users"."id" = $1 LIMIT $2  [["id", 9], ["LIMIT", 1]]
  User Load (0.4ms)  SELECT  "users".* FROM "users" WHERE "users"."id" = $1 LIMIT $2  [["id", 6], ["LIMIT", 1]]
  CACHE User Load (0.0ms)  SELECT  "users".* FROM "users" WHERE "users"."id" = $1 LIMIT $2  [["id", 6], ["LIMIT", 1]]
  User Load (0.4ms)  SELECT  "users".* FROM "users" WHERE "users"."id" = $1 LIMIT $2  [["id", 13], ["LIMIT", 1]]
  CACHE User Load (0.0ms)  SELECT  "users".* FROM "users" WHERE "users"."id" = $1 LIMIT $2  [["id", 13], ["LIMIT", 1]]
  CACHE User Load (0.0ms)  SELECT  "users".* FROM "users" WHERE "users"."id" = $1 LIMIT $2  [["id", 13], ["LIMIT", 1]]
  CACHE User Load (0.0ms)  SELECT  "users".* FROM "users" WHERE "users"."id" = $1 LIMIT $2  [["id", 13], ["LIMIT", 1]]
  CACHE User Load (0.0ms)  SELECT  "users".* FROM "users" WHERE "users"."id" = $1 LIMIT $2  [["id", 13], ["LIMIT", 1]]
```

it even signals the N+1 in the logs, thanks to the [bullet](https://rubygems.org/gems/bullet) gem:

```text
GET /api/comments?a_id=13
USE eager loading detected
  Comment => [:user]
  Add to your finder: :includes => [:user]
```

So I think for `index` the best compromise is still preloading. Unfortunately I can't find a way to use, directly from Rails, `select` on the preloaded query, in our case the `SELECT users`. 

Rails AFAIK doesn't support select on preloaded relationships, see this comment by a member of the Rails team https://github.com/rails/rails/issues/15185#issuecomment-351868335 - The reason being how `includes` can switch from preloading to eager loading depending on the algorithm.

I was somehow able, thanks to a StackOverflow answer to come up with the following:

```ruby
article = Article.find 13
ActiveRecord::Associations::Preloader.new.preload(article.comments.select(%i[id processed_html user_id ancestry]), :user, select: %i[id name username]).first.send(:owners)
```

but: the API is not public, it forces us to use preload in all cases and it returns an array instead of a `ActiveRecord::Relation`, which breaks `.arrange` to create the tree back. There's also a gem, [rails_select_on_includes](https://rubygems.org/gems/rails_select_on_includes) (discouraged by the aforamentioned comment), but other than being a hack, it doesn't work with nested relations, which is what we need in thi case.

The only other way I can think of is to cheat, with something like:

```ruby
@users = User.where(id: article.comments.pluck(:user_id)).select(%i[id name username twitter_username github_username website_url])
@comments = article.comments.select(%i[id processed_html user_id ancestry]).arrange
```

and then in the view code we pass along the `@users` and when it comes to render the user object we do this:

```ruby
user = @users.find { |u| u.id == comment.user_id }
json.partial! "comment", comment: comment, user: user
```

I made it work locally. The queries are:

```sql
Processing by Api::V0::CommentsController#index as JSON
  Parameters: {"a_id"=>"13"}
  Article Load (0.6ms)  SELECT  "articles".* FROM "articles" WHERE "articles"."id" = $1 LIMIT $2  [["id", 13], ["LIMIT", 1]]
   (0.4ms)  SELECT "comments"."user_id" FROM "comments" WHERE "comments"."commentable_id" = $1 AND "comments"."commentable_type" = $2  [["commentable_id", 13], ["commentable_type", "Article"]]
  Comment Load (0.3ms)  SELECT "comments"."id", "comments"."processed_html", "comments"."user_id", "comments"."ancestry" FROM "comments" WHERE "comments"."commentable_id" = $1 AND "comments"."commentable_type" = $2  [["commentable_id", 13], ["commentable_type", "Article"]]
  User Load (0.7ms)  SELECT "users"."id", "users"."name", "users"."username", "users"."twitter_username", "users"."github_username", "users"."website_url" FROM "users" WHERE "users"."id" IN ($1, $2, $3, $4, $5, $6, $7, $8, $9)  [["id", 6], ["id", 9], ["id", 6], ["id", 3], ["id", 13], ["id", 13], ["id", 13], ["id", 13], ["id", 13]]
```

We have one more query (to get all the user ids to from comments), but we do indeed have smaller user objects in memory. Let me know what you think.

It would work in a similar way for the `show` action:

```sql
Processing by Api::V0::CommentsController#show as JSON
  Parameters: {"id"=>"1a"}
  Comment Load (0.2ms)  SELECT  "comments".* FROM "comments" WHERE "comments"."id" = $1 LIMIT $2  [["id", 36], ["LIMIT", 1]]
   (0.4ms)  SELECT "comments"."user_id" FROM "comments" WHERE (("comments"."ancestry" LIKE '6/35/36/%' OR "comments"."ancestry" = '6/35/36') OR "comments"."id" = 36)
  Comment Load (0.6ms)  SELECT "comments"."id", "comments"."processed_html", "comments"."user_id", "comments"."ancestry" FROM "comments" WHERE (("comments"."ancestry" LIKE '6/35/36/%' OR "comments"."ancestry" = '6/35/36') OR "comments"."id" = 36)
  User Load (0.8ms)  SELECT "users"."id", "users"."name", "users"."username", "users"."twitter_username", "users"."github_username", "users"."website_url" FROM "users" WHERE "users"."id" IN ($1, $2, $3, $4)  [["id", 13], ["id", 13], ["id", 13], ["id", 13]]
```

we would have one more query but can build a custom user object

Reply:
Hmmm I think the hack might be worth it just knowing how big our User objects are and that if we load a lot of them we are going to load a lot of data. What do you both think? I'm game for either way I do worry about memory though a bit with the initial approach.

Reply:
It's hard to make a decision about those two without testing it. 

On one side we have "3 queries and more memory" and on the other we have "4 queries and less memory than option 1". Instictively I would lean on the second BUT:

- the API is currently broken which allows us to have a little bit of wiggle room in testing
- we have edge caching
- we can test the current one, benchmark it with tools like `wrk` or `hey` by requesting an article with lots of comment, we can literally find in the DB the article that has the most
- while the benchmark goes we can check what happens to the memory
- if we are not satisfied we test the second one redoing the benchmark
- if both are too slow or too resource intensive we add an additional layer of caching, hopefully we don't need it

Either way, I would very much like to benchmark how this endpoint performs

Reply:
Lets leave it how it is and we can address it later if performance becomes an issue. No need to optimize prematurely